src/xwrap.rs

```diff
@@ -154,91 +188,92 @@ impl Drop for Display {
 }
 
 impl Image {
-    fn from_raw_ximage(ximage: *mut xlib::XImage) -> Option<Image> {
-        if ximage.is_null() {
-            None
-        } else {
-            Some(Image { handle: ximage })
-        }
-    }
-
     pub fn to_image_buffer(&self) -> Option<RgbaImage> {
-        let img = unsafe { &*self.handle };
-
-        if (img.red_mask, img.green_mask, img.blue_mask) == (0xF800, 0x07E0, 0x001F) {
+        if (
+            self.visual.red_mask,
+            self.visual.green_mask,
+            self.visual.blue_mask,
+        ) == (0xF800, 0x07E0, 0x001F)
+        {
             return self.to_image_buffer_rgb565();
         }
 
-        let bytes_per_pixel = match (img.depth, img.bits_per_pixel) {
+        let bytes_per_pixel = match (self.format.depth, self.format.bits_per_pixel) {
             (24, 24) => 3,
             (24, 32) | (32, 32) => 4,
             _ => return None,
         };
 
+        let pad = match self.format.scanline_pad {
+            32 => 4,
+            _ => return None,
+        };
+        let bytes_per_line = (self.w * bytes_per_pixel + pad - 1) / pad * pad;
+
         // Compute subpixel offsets into each pixel according the the bitmasks X gives us
         // Only 8 bit, byte-aligned values are supported
         // Truncate masks to the lower 32 bits as that is the maximum pixel size
         macro_rules! channel_offset {
-            ($mask:expr) => {
-                match (img.byte_order, $mask & 0xFFFFFFFF) {
-                    (0, 0xFF) | (1, 0xFF000000) => 0,
-                    (0, 0xFF00) | (1, 0xFF0000) => 1,
-                    (0, 0xFF0000) | (1, 0xFF00) => 2,
-                    (0, 0xFF000000) | (1, 0xFF) => 3,
+            ($mask:expr) => {{
+                use xproto::ImageOrder as O;
+                match (self.byte_order, $mask & 0xFFFFFFFF) {
+                    (O::LSB_FIRST, 0xFF) | (O::MSB_FIRST, 0xFEFFFFFF) => 0,
```

Comment:
Actually, I do have a comment. This `FEF` looks wrong (why `E`) and this whole constant seems off, too. I think this is meant to be `0xFF000000`.

Reply:
Good catch. Looks like an editing mishap since the original code was correct.